truffle/src/main/ruby/core/type.rb

```diff
@@ -210,6 +220,64 @@ def self.execute_check_convert_type(obj, cls, meth)
       raise TypeError, msg
     end
 
+    def self.rb_check_convert_type(obj, cls, meth)
+      return obj if object_kind_of?(obj, cls)
+      v = convert_type(obj, cls, meth, false)
+      return nil if v.nil?
+      unless object_kind_of?(v, cls)
+        raise TypeError, "Coercion error: obj.#{meth} did NOT return a #{cls} (was #{object_class(v)})"
+      end
+      v
+    end
+
+    def self.convert_type(obj, cls, meth, raise_on_error)
+      r = check_funcall(obj, meth)
+      if undefined.equal?(r)
+        if raise_on_error
+          raise TypeError, "can't convert #{obj} into #{cls} with #{meth}"
+        end
+        return nil
+      end
+      r
+    end
+
+    def self.check_funcall(recv, meth, args = undefined)
+      check_funcall_default(recv, meth, args, undefined)
+    end
+
+    def self.check_funcall_default(recv, meth, args, default)
+      return default unless check_funcall_respond_to(recv, meth, true)
+      unless check_funcall_callable(recv, meth)
+        return check_funcall_missing(recv, meth, args, true, default);
+      end
+      recv.__send__(meth)
+    end
+
+    def self.check_funcall_respond_to(obj, meth, priv)
+      # TODO Review BJF vm_respond_to
+      return true unless object_respond_to_no_built_in?(obj, :respond_to?, true)
+      !!obj.__send__(:respond_to?, meth, true)
+    end
+
+    def self.check_funcall_missing(recv, meth, args, respond, default)
+      ret = basic_obj_respond_to_missing(recv, meth, false) #PRIV false
+      return default unless undefined.equal?(ret) || ret
+      if object_respond_to_no_built_in?(recv, :method_missing, true)
+        begin
+          return recv.__send__(:method_missing, meth)    # , args
+        rescue NoMethodError
+          # TODO BJF usually more is done here
+          raise
```

Comment:
@eregon I think you right, I've added a new commit with more `check_funcall_missing` implemented which resolves the spec failures